transformer_engine/jax/transformer.py

```diff
@@ -0,0 +1,911 @@
+                num_heads=self.num_attention_heads,
+                dtype=self.dtype,
+                head_dim=head_dim,
+                transpose_batch_sequence=self.transpose_batch_sequence,
+                dropout_rate=self.attention_dropout,
+                dropout_rng_name=self.dropout_rng_name,
+                layernorm_type=self.layernorm_type,
+                layernorm_epsilon=self.layernorm_epsilon,
+                apply_residual_connection_post_layernorm=self.
+                apply_residual_connection_post_layernorm,
+                output_layernorm=False,    # Must do LayerNorm before MHA.
+                attn_type=AttentionType.PADDING,
+                float32_logits=self.float32_attention_logits,
+                scale_attn_logits=self.scale_attn_logits,
+                scaled_query_init=self.scaled_query_init,
+                fuse_qkv=self.fuse_qkv_params,
+                kernel_init=self.mha_kernel_init,
+                use_bias=self.use_bias,
+                bias_init=self.bias_init,
+                name='encoder_decoder_attention')(x,
+                                                  encoded,
+                                                  encoder_decoder_mask,
+                                                  deterministic=deterministic)
+            y = nn.Dropout(rate=self.hidden_dropout,
+                           broadcast_dims=(-2,))(y, deterministic=deterministic)
+            mlp_input = y + residual
+
+        # MlpBlock
+        residual = mlp_input
+        z, ln_out = LayerNormMlpBlock(
+            layernorm_type=self.layernorm_type,
+            epsilon=self.layernorm_epsilon,
+            major_sharding_type=infer_major_sharding_type(),
+            transpose_batch_sequence=self.transpose_batch_sequence,
+            return_layernorm_output=self.apply_residual_connection_post_layernorm,
+            intermediate_dim=self.mlp_hidden_size,
+            activations=self.mlp_activations,
+            intermediate_dropout_rate=self.hidden_dropout,
+            dtype=self.dtype,
+            scale_axes=('embed',),
+            kernel_init=self.mlp_kernel_init,
+            kernel_axes_1=('embed', 'act', 'mlp'),
+            kernel_axes_2=('mlp', 'embed'),
+            use_bias=self.use_bias,
+            bias_init=self.bias_init,
+            name='mlp',
+        )(mlp_input, deterministic=deterministic)
+
+        if self.apply_residual_connection_post_layernorm:
+            assert ln_out is not None
+            residual = ln_out
+
+        z = nn.Dropout(rate=self.hidden_dropout, broadcast_dims=(-2,))(z,
+                                                                       deterministic=deterministic)
+        if self.drop_path > 0.0:
+            drop_path_shape = _generate_drop_path_shape(z.shape, batch_dim)
+            z = nn.Dropout(rate=self.drop_path,
+                           broadcast_dims=drop_path_shape)(z, deterministic=deterministic)
```

Comment:
The two dropouts apply random sample for different dimensions. The key is second argument `broadcast_dims`. TE/PyT has similar behavior: https://github.com/NVIDIA/TransformerEngine/blob/main/transformer_engine/pytorch/transformer.py#L1201-L1212

Reply:
In [transformer_engine/pytorch/transformer.py#L1201-L1212](https://github.com/NVIDIA/TransformerEngine/blob/main/transformer_engine/pytorch/transformer.py#L1201-L1212), the `self.drop_path` is also a dropout op.
```python  
      if self.drop_path is None:
            with self.bias_dropout_add_exec_handler():
                bda_output = bias_dropout_add_func(
                    attention_output, attention_bias, residual, self.hidden_dropout
                )
        else:
            out = torch.nn.functional.dropout(
                attention_output + attention_bias,
                p=self.hidden_dropout,
                training=self.training,
            )
            bda_output = residual + self.drop_path(out)
```

Reply:
AFAIK, it is a misalignment between JAX and PyTorch.

In TE/JAX, the input shape is (batch, seqlen, hidden). We create a dropout mask with shape (batch, 1, hidden) and apply (broadcast) it to the input tensor. Thus, different seqlen share the same dropout mask. We assume that the last 2 dimensions are always seqlen and hidden, so broadcast_dims=-2. This logic is designed with reference to the [T5X](https://github.com/google-research/t5x) model.

On the contrary, it looks like PyTorch doesn’t provide such a feature.

@timmoon10 ,
We can add an option to choose whether the dropout behavior should follow TE/PyT or T5X settings.
Is it what you are looking for?

Reply:
Hmm, after discussion with team member, T5X needs broadcast_dims but broadcast_dims=-2 is not really correct.
For perf reason, we might transpose batch_size. In that case, -2 is incorrect. The new commit - 75f9933 will fix this issue.

Reply:
I see, the new commit makes things clearer. I think adding an option for the dropout broadcasting is the right approach. It'll be needed if anyone tries implementing GPT in JAX. In the future we'll also want to add this option in PyTorch in case anyone wants to implement T5.

Reply:
@timmoon10 ,
The commit [94c7d37](https://github.com/NVIDIA/TransformerEngine/pull/54/commits/94c7d37e5243377b828f41404c7644193c0e3f6c) aligns the `hidden dropout mask` between TE/PyTorch and TE/JAX as default